src/jit/lower.cpp

```diff
@@ -3841,46 +3841,266 @@ GenTree* Lowering::LowerAdd(GenTree* node)
 }
 
 //------------------------------------------------------------------------
-// LowerUnsignedDivOrMod: transform GT_UDIV/GT_UMOD nodes with a const power of 2
-// divisor into GT_RSZ/GT_AND nodes.
+// GetUnsignedMagicNumberForDivide: Generates a magic number and shift amount for
+// the magic number unsigned division optimization.
 //
 // Arguments:
-//    node - pointer to the GT_UDIV/GT_UMOD node to be lowered
+//    denom - The denominator
+//    shift - Pointer to the shift value to be returned
+//    add   - Pointer to a flag indicating the kind of code to generate
 //
-void Lowering::LowerUnsignedDivOrMod(GenTree* node)
+// Returns:
+//    The magic number.
+//
+// Notes:
+//    This code is previously from UTC where it notes it was taken from
+//   _The_PowerPC_Compiler_Writer's_Guide_, pages 57-58. The paper is is based on
+//   is "Division by invariant integers using multiplication" by Torbjorn Granlund
+//   and Peter L. Montgomery in PLDI 94
+
+template <typename U>
+U GetUnsignedMagicNumberForDivide(U denom, int* shift /*out*/, bool* add /*out*/)
```

Comment:
@Tornhoof Yes, it's the original algorithm. I've seen "Improved Division by Invariant Integers" but it seems that it skewed towards multi word division so I didn't pay too much attention to it. Also, one of its basic premises is that that MULHI is slightly slower than a normal MUL but that doesn't seem to be the case anymore on current CPUs. Granlund's own instruction timing tables show that the latency dropped from 10 cycles (on Nehalem) to 3 cycles (on Haswell).

I was actually looking at another potential improvement: http://ridiculousfish.com/files/faster_unsigned_division_by_constants.pdf. It simplifies a bit the adjustment code required by stubborn divisors like 7. But I haven't seen any compiler using that and I have no idea if it's actually correct.